src/finch/codegen/numba_backend.py

```diff
@@ -378,3 +509,128 @@ def numba_repack(self, ctx, lhs, rhs):
         to the original object to update.
         """
         ...
+
+
+def serialize_struct_to_numba(fmt: AssemblyStructFormat, obj) -> Any:
+    args = [getattr(obj, name) for (name, _) in fmt.struct_fields]
+    return numba_type(fmt)(*args)
+
+
+register_property(
+    AssemblyStructFormat, "serialize_to_numba", "__attr__", serialize_struct_to_numba
+)
+
+
+def deserialize_struct_from_numba(
+    fmt: AssemblyStructFormat, obj, numba_struct: Any
+) -> None:
+    if fmt.is_mutable:
+        for name in fmt.struct_fieldnames:
+            setattr(obj, name, getattr(numba_struct, name))
+        return
+
+
+register_property(
+    AssemblyStructFormat,
+    "deserialize_from_numba",
+    "__attr__",
+    deserialize_struct_from_numba,
+)
+
+numba_structs: dict[Any, Any] = {}
+numba_structnames = Namespace()
+
+
+def really_numba_type(x):
+    if issubclass(x, np.generic):
+        return numba.from_dtype(x)
+    return numba.extending.as_numba_type(x)
+
+
+def struct_numba_type(fmt: AssemblyStructFormat):
+    res = numba_structs.get(fmt)
+    if res:
+        return res
+    spec = [
+        (name, really_numba_type(numba_type(field_type)))
+        for (name, field_type) in fmt.struct_fields
+    ]
+    class_name = numba_structnames.freshen("Numba", fmt.struct_name)
+    # Dynamically define __init__ based on spec, unrolling the arguments
+    field_names = [name for name, _ in spec]
+    # Build the argument list for __init__
+    arg_list = ", ".join(field_names)
+    # Build the body of __init__ to assign each argument to self
+    body_lines = [f"        self.{name} = {name}" for name in field_names]
+    body = "\n".join(body_lines)
+    # Compose the full class source
+    init_func_src = f"def __init__(self, {arg_list}):\n{body if body else '    pass'}"
+    name_func_src = (
+        f"@staticmethod\n    def numba_name():\n        return '{class_name}'"
+    )
```

Comment:
[`textwrap`](https://docs.python.org/3/library/textwrap.html) might be useful to make this more readable.

Reply:
Done! Now it should be readable.